src/java.base/share/classes/jdk/internal/classfile/components/ClassRemapper.java

```diff
@@ -0,0 +1,401 @@
+ * <p>
+ * The re-mapping is applied to superclass, interfaces, all kinds of descriptors and signatures,
+ * all attributes referencing classes in any form (including all types of annotations),
+ * and to all instructions referencing to classes.
+ * <p>
+ * Primitive types and arrays are never subjects of mapping and are not allowed targets of mapping.
+ * <p>
+ * Arrays of reference types are always decomposed, mapped as the base reference types and composed back to arrays.
+ */
+public sealed interface ClassRemapper extends ClassTransform {
+
+    /**
+     * Creates new instance of ClassRemapper instructed with a class map.
+     * Map may contain only re-mapping entries, identity mapping is applied by default.
+     * @param classMap class map
+     * @return new instance of ClassRemapper
+     */
+    static ClassRemapper of(Map<ClassDesc, ClassDesc> classMap) {
+        return of(desc -> classMap.getOrDefault(desc, desc));
+    }
+
+    /**
+     * Creates new instance of ClassRemapper instructed with a map function.
+     * Map function must return valid {@link java.lang.constant.ClassDesc} of an interface
+     * or a class, even for identity mappings.
+     * @param mapFunction class map function
+     * @return new instance of ClassRemapper
+     */
+    static ClassRemapper of(Function<ClassDesc, ClassDesc> mapFunction) {
+        return new ClassRemapperImpl(mapFunction);
+    }
+
+    /**
+     * Access method to internal class mapping function.
+     * @param desc source class
+     * @return class target class
+     */
+    ClassDesc map(ClassDesc desc);
+
+    /**
+     * Returns this ClassRemapper as {@link jdk.internal.classfile.FieldTransform} instance
+     * @return this ClassRemapper as {@link jdk.internal.classfile.FieldTransform} instance
+     */
+    FieldTransform asFieldTransform();
+
+    /**
+     * Returns this ClassRemapper as {@link jdk.internal.classfile.MethodTransform} instance
+     * @return this ClassRemapper as {@link jdk.internal.classfile.MethodTransform} instance
+     */
+    MethodTransform asMethodTransform();
+
+    /**
+     * Returns this ClassRemapper as {@link jdk.internal.classfile.CodeTransform} instance
+     * @return this ClassRemapper as {@link jdk.internal.classfile.CodeTransform} instance
+     */
+    CodeTransform asCodeTransform();
+
+    /**
+     * Remaps the whole ClassModel into a new class file, including the class name.
+     * @param clm class model to re-map
+     * @return re-mapped class file bytes
+     */
+    default byte[] remapClass(ClassModel clm) {
+        return Classfile.build(map(clm.thisClass().asSymbol()),
+                clb -> clm.forEachElement(resolve(clb).consumer()));
+    }
+
+    record ClassRemapperImpl(Function<ClassDesc, ClassDesc> mapFunction) implements ClassRemapper {
+
+        @Override
+        public void accept(ClassBuilder clb, ClassElement cle) {
+            switch (cle) {
+                case FieldModel fm ->
```

Comment:
What about NestMembers, NestHost, PermittedSubclasses (and probably others) ?

Reply:
Right, ClassRemapper is incomplete, will work on it.

Reply:
fixed